### PR TITLE
Reduce concurrency issues when consuming tasks

### PR DESCRIPTION
**Pull request by thc202**
Track tasks being consumed in `UnfiredFragmentCandidates` instead of `CrawlTaskConsumer`, to have the changes in number of consumers and pending states done atomically.
Also, do not put back into the queue the state being consumed otherwise it would allow several tasks to try consume the same state at the same time.

### CODE REVIEW

**Comment by thc202** [comment]
This fixes cases where Crawljax would exit early with exhausted reason while there were still consumers running (this is more noticeable when using a higher number of browsers).

**Comment by thc202** [comment]
I'm investigating the test failures.

File: core/src/main/java/com/crawljax/core/UnfiredFragmentCandidates.java [hunk 1]
```diff
@@ -305,33 +323,54 @@ void addActions(List<CandidateCrawlAction> actions, StateVertex state) {
...
+                    pendingStates,
+                    statesWithCandidates);
+            return empty;
+        } finally {
+            consumersReadLock.unlock();
+        }
     }
 
     /**
      * @param fragmentManager
```
[the remark this answers is not in the document]
Comment (thc202) [reply core/src/main/java/com/crawljax/core/UnfiredFragmentCandidates.java]: This is existing JavaDoc.

File: core/src/main/java/com/crawljax/core/UnfiredFragmentCandidates.java [hunk 2]
```diff
@@ -305,33 +323,54 @@ void addActions(List<CandidateCrawlAction> actions, StateVertex state) {
...
+    private void addPendingState(StateVertex state) {
+        pendingStates++;
+        statesWithCandidates.add(state.getId());
+        LOG.info("There are {} states with unfired actions", pendingStates);
     }
 
     /**
-     * @return If there are any pending actions to be crawled. This method is not threadsafe and might
-     * return a stale value.
+     * @return If there are any pending actions to be crawled (and no state is being crawled).
```
[the remark this answers is not in the document]
Comment (thc202) [reply core/src/main/java/com/crawljax/core/UnfiredFragmentCandidates.java:345]: This is existing JavaDoc, the text was reworded to match the current expectations.

File: core/src/main/java/com/crawljax/core/UnfiredFragmentCandidates.java [hunk 3]
```diff
@@ -369,14 +406,71 @@ public StateVertex awaitNewTask(
...
+        }
+
+        consumersWriteLock.lock();
+        try {
+            runningConsumers--;
+
+            int stateId = state.getId();
+            Lock lock = locks.get(stateId);
+            try {
+                lock.lock();
```
[the remark this answers is not in the document]
Comment (thc202) [reply core/src/main/java/com/crawljax/core/UnfiredFragmentCandidates.java:453]: This is how it's being done in other places.

**Comment by thc202** [comment]
CI is green again.

**Comment by thc202** [comment]
I will do further integration tests in ZAP, to confirm this properly addresses the issue.

**Review by amesbah — APPROVED** [review]

**Pull request merged**